src/tools/rbd_cache/SimplePolicy.hpp

```diff
@@ -3,178 +3,158 @@
 
 #include "Policy.hpp"
 #include "include/lru.h"
+#include "common/RWLock.h"
 #include "common/Mutex.h"
 
 #include <vector>
 #include <unordered_map>
 #include <string>
 
+namespace rbd {
+namespace cache {
+
+
 class SimplePolicy : public Policy {
 public:
-  SimplePolicy(uint64_t block_num, float level)
-    : m_level(level),
-      m_lock("SimplePolicy"),
-      m_entry_count(block_num)
+  SimplePolicy(uint64_t block_num, float watermark)
+    : m_watermark(watermark), m_entry_count(block_num),
+      m_cache_map_lock("rbd::cache::SimplePolicy::m_cache_map_lock"),
+      m_free_list_lock("rbd::cache::SimplePolicy::m_free_list_lock")
   {
 
-    Entry m_entries[m_entry_count];
-
-    for(auto &entry : m_entries) {
-      m_free_lru.lru_insert_bot(&entry);
+    for(uint64_t i = 0; i < m_entry_count; i++) {
+      m_free_list.push_back(new Entry());
     }
+
   }
 
-  ~SimplePolicy() {}
+  ~SimplePolicy() {
+    for(uint64_t i = 0; i < m_entry_count; i++) {
+      Entry* entry = reinterpret_cast<Entry*>(m_free_list.front());
+      delete entry;
+      m_free_list.pop_front();
+    }
+  }
 
   CACHESTATUS lookup_object(std::string cache_file_name) {
-    Mutex::Locker locker(m_lock);
 
-    auto entry_it = m_oid_to_entry.find(cache_file_name);
-    if(entry_it == m_oid_to_entry.end()) {
-      return NONE;
+    //TODO(): check race condition
+    RWLock::WLocker wlocker(m_cache_map_lock);
+
+    auto entry_it = m_cache_map.find(cache_file_name);
+    if(entry_it == m_cache_map.end()) {
+      Mutex::Locker locker(m_free_list_lock);
+      Entry* entry = reinterpret_cast<Entry*>(m_free_list.front());
+      assert(entry != nullptr);
```

Comment:
should it be `assert(m_free_list.size())`?

Reply:
yes, you are right. Calling "deque::front()" on an empty container causes undefined behavior.